### ISSUE TITLE
RTG loss and CTG loss become NaN

### ISSUE THREAD

**Issue by 729248399** [HenryLHH/fusion#2]
Hi, thank you for sharing this great work and the open-source implementation!
I encountered an issue when running train_fusion.py. During training, both the RTG loss and CTG loss become NaN, while other loss terms appear normal. I’m wondering if you have seen a similar issue before or if there are any known causes or fixes.
Here are some of the training logs for reference:
num of updates: 500
action loss: 0.05633
rtg loss: nan
ctg loss: nan
dynamics loss: 0.05386
eval avg reward: 24.12366
eval avg ep len: 97.34000
eval avg succ: 0.00000
eval avg crash: 0.00000
eval avg oor: 1.00000
eval avg overtime: 0.00000
eval avg cost: 0.00000
I would really appreciate any suggestions or insights. Thanks in advance!

**Comment by HenryLHH** [comment]
Hey there, sorry for the late response. Are you using the same rollout data attached via the Google Drive link, or your custom rollout data from MetaDrive for your training job? 
If i remember correctly, when using different datasets/rollout environments for dataset rollout, sometimes it may run into this issue due to some numerical difference of reward/cost scale, and may be fixed by re-normalization of the GT labels of rewards and costs. Let me know if you have fixed this issue.

**Comment by 729248399** [comment]
> Hey there, sorry for the late response. Are you using the same rollout data attached via the Google Drive link, or your custom rollout data from MetaDrive for your training job? If i remember correctly, when using different datasets/rollout environments for dataset rollout, sometimes it may run into this issue due to some numerical difference of reward/cost scale, and may be fixed by re-normalization of the GT labels of rewards and costs. Let me know if you have fixed this issue.

Hi, thanks for getting back to me.I’m using the original dataset provided through the Google Drive link.
I found that the issue seems to stem from the loss computation code. In the original implementation, after masking and reshaping, the variables return_preds, rtg_target, returns_preds_cost, and ctg_target all have a shape of [N, 1]. However, the code calculates the losses using these values.
rtg_loss = F.mse_loss(rtg_target[:, 1:], return_preds[:, :-1])
ctg_loss = F.mse_loss(ctg_target[:, 1:], returns_preds_cost[:, :-1])
Since the tensors have only one feature dimension, slicing results in empty tensors with the shape [N, 0]. This can lead to NaN values in the loss calculations during training. 
I also noticed the state prediction loss: it seems that slicing is performed along the feature dimension rather than the time dimension.
After fixing the issue with dimension handling in the loss calculation, the NaN values disappeared.
Please let me know if I’m missing anything.

**Comment by HenryLHH** [comment]
Ah nice catch! I think I have fixed it in my experiments but forgot to push it to this branch. Would you mind putting up a PR for your fix? Thanks for raising this again!

**Comment by 729248399** [comment]
Thanks for confirming!
I noticed that some evaluation metrics (e.g., success rate) are slightly lower than those reported in the paper. 
Because of that, I’m not completely sure whether my modification matches the exact version used in the paper’s experiments.
If you still have access to the code version that was actually used to obtain the paper’s results, that would be extremely helpful for reproducing those results. 
Thanks again for the clarification!
